src/store/reducers/messagesReducer.ts

```diff
@@ -1,30 +1,39 @@
-import { List } from 'immutable';
+import { MessagesState } from '../types';
 
 import { createReducer } from '../../utils/createReducer';
 import { createNewMessage, createLinkSnippet, createComponentMessage } from '../../utils/messages';
 import { MESSAGE_SENDER } from '../../constants';
-
-import * as actionTypes from '../actions/actionTypes';
-
-const initialState = List([]);
+import {
+  MessagesActions,
+  ADD_NEW_USER_MESSAGE,
+  ADD_NEW_RESPONSE_MESSAGE,
+  ADD_NEW_LINK_SNIPPET,
+  ADD_COMPONENT_MESSAGE,
+  DROP_MESSAGES,
+  HIDE_AVATAR
+} from '../actions/types';
+
+const initialState = {
+  messages: []
+};
 
 const messagesReducer = {
-  [actionTypes.ADD_NEW_USER_MESSAGE]: (state, { text }) =>
-    state.push(createNewMessage(text, MESSAGE_SENDER.CLIENT)),
+  [ADD_NEW_USER_MESSAGE]: (state: MessagesState, { text }) =>
+    ({ messages: [...state.messages, createNewMessage(text, MESSAGE_SENDER.CLIENT)] }),
 
-  [actionTypes.ADD_NEW_RESPONSE_MESSAGE]: (state, { text }) =>
-    state.push(createNewMessage(text, MESSAGE_SENDER.RESPONSE)),
+  [ADD_NEW_RESPONSE_MESSAGE]: (state: MessagesState, { text }) =>
+    ({ messages: [...state.messages, createNewMessage(text, MESSAGE_SENDER.RESPONSE)] }),
 
-  [actionTypes.ADD_NEW_LINK_SNIPPET]: (state, { link }) =>
-    state.push(createLinkSnippet(link)),
+  [ADD_NEW_LINK_SNIPPET]: (state: MessagesState, { link }) =>
+    ({ messages: [...state.messages, createLinkSnippet(link)] }),
 
-  [actionTypes.ADD_COMPONENT_MESSAGE]: (state, { component, props, showAvatar }) =>
-    state.push(createComponentMessage(component, props, showAvatar)),
+  [ADD_COMPONENT_MESSAGE]: (state: MessagesState, { component, props, showAvatar }) =>
+    ({ messages: [...state.messages, createComponentMessage(component, props, showAvatar)] }),
 
-  [actionTypes.DROP_MESSAGES]: () => List([]),
+  [DROP_MESSAGES]: () => ({ messages: [] }),
 
-  [actionTypes.HIDE_AVATAR]: (state, { index }) =>
-    state.update(index, message => message.set('showAvatar', false))
+  [HIDE_AVATAR]: (state: MessagesState, { index }) =>
+    state.messages[index].showAvatar = false
```

Comment:
state should be immutable

Reply:
I will add `seamless-immutable` in another PR

Reply:
Still, I don't think we should leave it like this. I think it should work with just this:
```
messages = [...state.messages];
messages[index].showsAvatar = false;
return { ...state, messages };
```